File: dump_tree.rb
#!/usr/bin/env ruby

class Node
  attr_accessor :value, :children

  def initialize(value, children=[])
    self.value = value
    self.children = children.is_a?(Array) ? children : [children]
  end
end

class Tree
  attr_accessor :root

  def initialize(root)
    self.root = root
  end

  def to_s
    Tree.to_string(root)
  end

  def print
    Tree.print_tree(root)
  end

  def self.to_string(node)
    memo = node.value + ": "
    node.children.each do |sub|
      memo += to_string(sub)
    end
    memo
  end

  def self.print_tree(node)
    print node.value + ": "
    node.children.each do |sub|
      print_tree(sub)
    end
  end
end

#        "ab"
#        /  \
#      "ef"  "gh"
#     / | \    \
#  "x" "y" "z"  "p"

root = Node.new("ab", [Node.new("ef", [
    Node.new("x"), Node.new("y"), Node.new("z")]),
    Node.new("gh", Node.new("p"))])

Tree.print_tree(root)
puts
puts Tree.to_string(root)
tree = Tree.new root
tree.print
puts
puts tree.to_s

__END__

ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:

File: dump_tree.js
#!/usr/bin/env node

"use strict";

var Iterator = (function () {
  function Iterator(node) {
    this.stack = [node];
  }
  Iterator.prototype.next = function () {
    if (this.hasNext()) {
      var node = this.stack.pop();
      for (let child of node.children.reverse()) {
        this.stack.push(child);
      }
      return node;
    } else {
      return null;
    }
  }
  Iterator.prototype.hasNext = function () {
    return (this.stack.length > 0);
  }
  return Iterator;
})();

var Node = (function () {
  function Node(value, children) {
    this.value = value;
    this.children = (typeof children === 'undefined') ? [] :
        Array.isArray(children) ? children : [children];
  }
  return Node;
})();

var Tree = (function () {
  function Tree(node) {
    this.root = root;
  };
  Tree.prototype.to_s = function () {
    return Tree.to_string(this.root);
  };
  Tree.prototype.print = function () {
    Tree.print_tree(this.root);
  };
  Tree.to_string = function (node) {
    var memo = node.value + ": "
    for (let sub of node.children) {
      memo += Tree.to_string(sub);
    }
    return memo;
  };
  Tree.print_tree = function (node) {
    process.stdout.write(node.value + ": ");
    for (let sub of node.children) {
      Tree.print_tree(sub);
    }
  };
  return Tree;
})();

//        "ab"
//        /  \
//      "ef"  "gh"
//     / | \    \
//  "x" "y" "z"  "p"

var root = new Node("ab", [new Node("ef", [
    new Node("x"), new Node("y"), new Node("z")]),
    new Node("gh", new Node("p"))])

console.log(Tree.to_string(root));
Tree.print_tree(root);
console.log('');
var tree = new Tree(root);
console.log(tree.to_s());
tree.print();
console.log('');

var iter = new Iterator(root);
while(iter.hasNext()) {
  process.stdout.write(iter.next().value + ": ");
}
console.log('');

/*

ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:
ab: ef: x: y: z: gh: p:

*/
